### PR TITLE
promql: ensure series in matrix values returned for instant queries are sorted

### PR DESCRIPTION
**Pull request by charleskorn**
The PromQL engine always sorts matrix values returned for range queries ([source](https://github.com/prometheus/prometheus/blob/3b8b57700c469c7cde84e1d8f9d383cb8fe11ab0/promql/engine.go#L794)), but does not do the same thing for matrix values returned for instant queries.

This PR makes the behaviour for instant queries consistent with range queries: matrix values are now always sorted by series. 

I've added the `TestInstantQueryWithRangeVectorSelector/matches series with points in range` test case to demonstrate the issue and that it has been fixed. (Note that without the fix, the series order is unpredictable, so you may need to run the test multiple times to see the failure.)

I've also added some other test cases for instant queries that return matrices given there weren't any test cases for this scenario.

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -752,6 +752,7 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
 		case parser.ValueTypeScalar:
 			return Scalar{V: mat[0].Floats[0].F, T: start}, warnings, nil
 		case parser.ValueTypeMatrix:
+			sort.Sort(mat)
```
Comment (krajorama) [inline promql/engine.go:755]: Let's also add the `sortSpanTimer` from the range query to be consistent.

**Review by krajorama — COMMENTED** [review]
Looks good, one suggestion and I started prombench.

File: promql/engine.go (hunk 1 above)
Comment (charleskorn) [reply promql/engine.go]: Good catch, added in Good catch, added in 0e934db.

**Review by krajorama — APPROVED** [review]
LGTM

**Comment by machine424** [comment]
Thanks.
> (Note that without the fix, the series order is unpredictable, so you may need to run the test multiple times to see the failure.)
If the operation is cheap, we could run it multiple times in a loop to increase the likelihood of a change in order.

Taking a look at the benchmark dashboards, I can see some overhead in the `CPU`/`query latency` and `Avg query engine timings/s` dashboards as expected. However, I am unsure if this overhead is negligible.

**Comment by darshanime** [comment]
> I can see some overhead in the CPU/query latency and Avg query engine timings/s dashboards as expected. However, I am unsure if this overhead is negligible.

Running the benchmarks would help quantify the overhead?

**Review by bboreham — APPROVED** [review]
Do you think this behaviour is documented anywhere? 
I think it should be.  Anyway we can go ahead with the change.

**Comment by bboreham** [comment]
> Running the benchmarks

We did run prombench, although it doesn't have this kind of query.

I don't think there are any benchmarks which currently test the case of an instant query returning a matrix.  Would you like to write some?

**Comment by charleskorn** [comment]
> Do you think this behaviour is documented anywhere? I think it should be.

I couldn't find it documented anywhere. Perhaps [here](https://prometheus.io/docs/prometheus/latest/querying/api/#range-vectors) makes sense?

Should we also require instant vectors to be sorted by series as well? Seems like this would be good for the sake of consistency.

**Comment by bboreham** [comment]
> I couldn't find it documented anywhere. Perhaps [here](https://prometheus.io/docs/prometheus/latest/querying/api/#range-vectors) makes sense?

:+1:

> Should we also require instant vectors to be sorted by series as well?

The code takes care not to do that, since there are PromQL functions like `sort` which give a specific order.
Arguably it could do it where the expression has not supplied an order, but that seems complicated.

**Comment by bboreham** [comment]
Related: the code silently ignores any `sort` in the PromQL when returning a matrix.
In my mind I imagine some heartless Prometheus maintainer gloating "they'll figure it out eventually", but we could add an Annotation to be nice.

[when multiple points are returned, which set do you sort on?]

**Comment by charleskorn** [comment]
> Related: the code silently ignores any `sort` in the PromQL when returning a matrix. In my mind I imagine some heartless Prometheus maintainer gloating "they'll figure it out eventually", but we could add an Annotation to be nice.
> 
> [when multiple points are returned, which set do you sort on?]

I've created https://github.com/prometheus/prometheus/issues/14115 so we don't lose this.

**Comment by charleskorn** [comment]
I've resolved the merge conflicts. The CI failure seems to be an unrelated flake.

**Review by bboreham — APPROVED** [review]
LGTM, thanks!

**Pull request merged**